File: web/src/batcher.ts
// Message batching with priority support
// CRC: crc-MessageBatcher.md
// Spec: protocol.md

import { Message, UpdateMessage } from './protocol';

/** Priority levels for batching */
export enum Priority {
  High = 0,
  Medium = 1,
  Low = 2,
}

/**
 * Parse priority suffix from property name.
 * @param propertyName Property name, possibly with :high/:med/:low suffix
 * @returns Tuple of [baseName, priority]
 */
export function parsePrioritySuffix(propertyName: string): [string, Priority] {
  if (propertyName.endsWith(':high')) {
    return [propertyName.slice(0, -5), Priority.High];
  }
  if (propertyName.endsWith(':med')) {
    return [propertyName.slice(0, -4), Priority.Medium];
  }
  if (propertyName.endsWith(':low')) {
    return [propertyName.slice(0, -4), Priority.Low];
  }
  return [propertyName, Priority.Medium];
}

/** Pending change for a variable */
interface PendingChange {
  varId: number;
  value?: unknown;
  valuePriority: Priority;
  hasValue: boolean;
  properties: Map<string, string>;
  propPriorities: Map<string, Priority>;
}

/** Message batcher that groups updates by priority */
export class MessageBatcher {
  private pending: Map<number, PendingChange> = new Map();

  /** Get or create pending change for variable */
  private getOrCreate(varId: number): PendingChange {
    let pc = this.pending.get(varId);
    if (!pc) {
      pc = {
        varId,
        valuePriority: Priority.Medium,
        hasValue: false,
        properties: new Map(),
        propPriorities: new Map(),
      };
      this.pending.set(varId, pc);
    }
    return pc;
  }

  /** Queue a value change with priority */
  queueValue(varId: number, value: unknown, priority: Priority = Priority.Medium): void {
    const pc = this.getOrCreate(varId);
    pc.value = value;
    pc.valuePriority = priority;
    pc.hasValue = true;
  }

  /**
   * Queue a property change.
   * Property name can include priority suffix (e.g., "viewdefs:high")
   */
  queueProperty(varId: number, propertyName: string, value: string): void {
    const [baseName, priority] = parsePrioritySuffix(propertyName);
    const pc = this.getOrCreate(varId);
    pc.properties.set(baseName, value);
    pc.propPriorities.set(baseName, priority);
  }

  /** Queue multiple properties */
  queueProperties(varId: number, properties: Record<string, string>): void {
    for (const [name, value] of Object.entries(properties)) {
      this.queueProperty(varId, name, value);
    }
  }

  /** Check if any changes are pending */
  isEmpty(): boolean {
    return this.pending.size === 0;
  }

  /** Build and return batched messages, clearing pending state */
  flush(): Message[] {
    if (this.pending.size === 0) {
      return [];
    }

    interface BatchEntry {
      priority: Priority;
      message: Message;
    }

    const entries: BatchEntry[] = [];

    for (const pc of this.pending.values()) {
      // Group properties by priority
      const highProps: Record<string, string> = {};
      const medProps: Record<string, string> = {};
      const lowProps: Record<string, string> = {};

      for (const [name, value] of pc.properties) {
        const priority = pc.propPriorities.get(name) ?? Priority.Medium;
        switch (priority) {
          case Priority.High:
            highProps[name] = value;
            break;
          case Priority.Medium:
            medProps[name] = value;
            break;
          case Priority.Low:
            lowProps[name] = value;
            break;
        }
      }

      // Create messages for each priority level with content
      // High priority
      if (
        Object.keys(highProps).length > 0 ||
        (pc.hasValue && pc.valuePriority === Priority.High)
      ) {
        const msg = this.createUpdateMessage(
          pc.varId,
          pc.hasValue && pc.valuePriority === Priority.High,
          pc.value,
          highProps
        );
        if (msg) {
          entries.push({ priority: Priority.High, message: msg });
        }
      }

      // Medium priority
      if (
        Object.keys(medProps).length > 0 ||
        (pc.hasValue && pc.valuePriority === Priority.Medium)
      ) {
        const msg = this.createUpdateMessage(
          pc.varId,
          pc.hasValue && pc.valuePriority === Priority.Medium,
          pc.value,
          medProps
        );
        if (msg) {
          entries.push({ priority: Priority.Medium, message: msg });
        }
      }

      // Low priority
      if (
        Object.keys(lowProps).length > 0 ||
        (pc.hasValue && pc.valuePriority === Priority.Low)
      ) {
        const msg = this.createUpdateMessage(
          pc.varId,
          pc.hasValue && pc.valuePriority === Priority.Low,
          pc.value,
          lowProps
        );
        if (msg) {
          entries.push({ priority: Priority.Low, message: msg });
        }
      }
    }

    // Clear pending
    this.pending.clear();

    // Sort by priority (High=0 first, then Medium=1, then Low=2)
    entries.sort((a, b) => a.priority - b.priority);

    return entries.map((e) => e.message);
  }

  /** Create update message for a variable */
  private createUpdateMessage(
    varId: number,
    includeValue: boolean,
    value: unknown,
    properties: Record<string, string>
  ): Message | null {
    if (!includeValue && Object.keys(properties).length === 0) {
      return null;
    }

    const update: UpdateMessage = { varId };
    if (includeValue) {
      update.value = value;
    }
    if (Object.keys(properties).length > 0) {
      update.properties = properties;
    }

    return { type: 'update', data: update };
  }

  /** Flush and return as JSON string (array if multiple, object if single) */
  flushJSON(): string | null {
    const messages = this.flush();
    if (messages.length === 0) {
      return null;
    }
    if (messages.length === 1) {
      return JSON.stringify(messages[0]);
    }
    return JSON.stringify(messages);
  }
}

/**
 * Parse incoming batch (array or single message).
 * @param json JSON string that may be a single message or array
 * @returns Array of messages
 */
export function parseBatch(json: string): Message[] {
  const parsed = JSON.parse(json);
  if (Array.isArray(parsed)) {
    return parsed;
  }
  return [parsed];
}

File: internal/protocol/batcher.go
// CRC: crc-MessageBatcher.md
// Spec: protocol.md
package protocol

import (
	"encoding/json"
	"strings"
	"sync"
)

// Priority levels for batching
type Priority int

const (
	PriorityHigh   Priority = 0
	PriorityMedium Priority = 1
	PriorityLow    Priority = 2
)

// ParsePrioritySuffix extracts priority from property name suffix.
// Returns base property name and priority.
// Examples: "viewdefs:high" -> ("viewdefs", PriorityHigh)
//
//	"name" -> ("name", PriorityMedium)
func ParsePrioritySuffix(propertyName string) (string, Priority) {
	if strings.HasSuffix(propertyName, ":high") {
		return strings.TrimSuffix(propertyName, ":high"), PriorityHigh
	}
	if strings.HasSuffix(propertyName, ":med") {
		return strings.TrimSuffix(propertyName, ":med"), PriorityMedium
	}
	if strings.HasSuffix(propertyName, ":low") {
		return strings.TrimSuffix(propertyName, ":low"), PriorityLow
	}
	return propertyName, PriorityMedium
}

// PendingChange represents a queued change for a variable.
type PendingChange struct {
	VarID         int64
	Value         json.RawMessage
	ValuePriority Priority
	HasValue      bool
	Properties    map[string]string     // property name -> value
	PropPriorities map[string]Priority  // property name -> priority
}

// MessageBatcher batches protocol messages by priority.
type MessageBatcher struct {
	pending map[int64]*PendingChange
	mu      sync.Mutex
}

// NewMessageBatcher creates a new message batcher.
func NewMessageBatcher() *MessageBatcher {
	return &MessageBatcher{
		pending: make(map[int64]*PendingChange),
	}
}

// getOrCreate returns existing pending change or creates a new one.
func (b *MessageBatcher) getOrCreate(varID int64) *PendingChange {
	if pc, ok := b.pending[varID]; ok {
		return pc
	}
	pc := &PendingChange{
		VarID:          varID,
		ValuePriority:  PriorityMedium,
		Properties:     make(map[string]string),
		PropPriorities: make(map[string]Priority),
	}
	b.pending[varID] = pc
	return pc
}

// QueueValue queues a value change with the given priority.
func (b *MessageBatcher) QueueValue(varID int64, value json.RawMessage, priority Priority) {
	b.mu.Lock()
	defer b.mu.Unlock()

	pc := b.getOrCreate(varID)
	pc.Value = value
	pc.ValuePriority = priority
	pc.HasValue = true
}

// QueueProperty queues a property change.
// Property name can include priority suffix (e.g., "viewdefs:high").
func (b *MessageBatcher) QueueProperty(varID int64, propertyName, value string) {
	baseName, priority := ParsePrioritySuffix(propertyName)

	b.mu.Lock()
	defer b.mu.Unlock()

	pc := b.getOrCreate(varID)
	pc.Properties[baseName] = value
	pc.PropPriorities[baseName] = priority
}

// QueueProperties queues multiple property changes.
func (b *MessageBatcher) QueueProperties(varID int64, properties map[string]string) {
	for name, value := range properties {
		b.QueueProperty(varID, name, value)
	}
}

// IsEmpty returns true if no changes are pending.
func (b *MessageBatcher) IsEmpty() bool {
	b.mu.Lock()
	defer b.mu.Unlock()
	return len(b.pending) == 0
}

// batchEntry represents a single update in the batch.
type batchEntry struct {
	priority Priority
	message  *Message
}

// Flush builds and returns the batched messages, clearing pending state.
// Returns nil if no changes are pending.
func (b *MessageBatcher) Flush() []*Message {
	b.mu.Lock()
	defer b.mu.Unlock()

	if len(b.pending) == 0 {
		return nil
	}

	// Collect all entries
	var entries []batchEntry

	for _, pc := range b.pending {
		// Group properties by priority
		highProps := make(map[string]string)
		medProps := make(map[string]string)
		lowProps := make(map[string]string)

		for name, value := range pc.Properties {
			priority := pc.PropPriorities[name]
			switch priority {
			case PriorityHigh:
				highProps[name] = value
			case PriorityMedium:
				medProps[name] = value
			case PriorityLow:
				lowProps[name] = value
			}
		}

		// Create messages for each priority level that has content
		// High priority
		if len(highProps) > 0 || (pc.HasValue && pc.ValuePriority == PriorityHigh) {
			msg := b.createUpdateMessage(pc.VarID,
				pc.HasValue && pc.ValuePriority == PriorityHigh,
				pc.Value,
				highProps)
			if msg != nil {
				entries = append(entries, batchEntry{PriorityHigh, msg})
			}
		}

		// Medium priority
		if len(medProps) > 0 || (pc.HasValue && pc.ValuePriority == PriorityMedium) {
			msg := b.createUpdateMessage(pc.VarID,
				pc.HasValue && pc.ValuePriority == PriorityMedium,
				pc.Value,
				medProps)
			if msg != nil {
				entries = append(entries, batchEntry{PriorityMedium, msg})
			}
		}

		// Low priority
		if len(lowProps) > 0 || (pc.HasValue && pc.ValuePriority == PriorityLow) {
			msg := b.createUpdateMessage(pc.VarID,
				pc.HasValue && pc.ValuePriority == PriorityLow,
				pc.Value,
				lowProps)
			if msg != nil {
				entries = append(entries, batchEntry{PriorityLow, msg})
			}
		}
	}

	// Clear pending
	b.pending = make(map[int64]*PendingChange)

	// Sort by priority (stable sort to preserve order within priority)
	// Simple insertion sort since we have only 3 priority levels
	result := make([]*Message, 0, len(entries))
	for p := PriorityHigh; p <= PriorityLow; p++ {
		for _, e := range entries {
			if e.priority == p {
				result = append(result, e.message)
			}
		}
	}

	return result
}

// createUpdateMessage creates an update message for a variable.
func (b *MessageBatcher) createUpdateMessage(varID int64, includeValue bool, value json.RawMessage, properties map[string]string) *Message {
	// Skip if nothing to send
	if !includeValue && len(properties) == 0 {
		return nil
	}

	update := UpdateMessage{
		VarID: varID,
	}
	if includeValue {
		update.Value = value
	}
	if len(properties) > 0 {
		update.Properties = properties
	}

	msg, _ := NewMessage(MsgUpdate, update)
	return msg
}

// FlushJSON returns the batch as a JSON array or single message.
// Returns nil if no changes are pending.
func (b *MessageBatcher) FlushJSON() ([]byte, error) {
	messages := b.Flush()
	if messages == nil || len(messages) == 0 {
		return nil, nil
	}

	if len(messages) == 1 {
		return messages[0].Encode()
	}

	return json.Marshal(messages)
}
